File: histogra.py
def area(temp, stack):
    if len(stack) > 0:
        currentArea = arr[temp]*(i-stack[-1]-1)
    else:
        currentArea = arr[temp]*i
    return currentArea 


arr = map(int, raw_input().split())
n  = arr.pop(0)
while(n != 0):
   
    maxArea = 0
    currentArea = 0 
    stack = []

    i = 0 ;
    while(i < n):
        if len(stack) == 0:
            stack.append(i)
            i += 1 
        else:
            if arr[i] < arr[stack[-1]]:
                temp =stack.pop() 
                currentArea = area(temp , stack) 
                maxArea = max(maxArea, currentArea)
            else:
                stack.append(i)
                i += 1 

    # Calculate the area of the stack elements
    while(len(stack) > 0):
        temp = stack.pop()
        currentArea = area(temp, stack)
        maxArea = max(maxArea, currentArea)
    print maxArea
    arr = map(int, raw_input().split())
    n  = arr.pop(0)

File: histogra.cpp
#include<bits/stdc++.h>

using namespace std ;

stack<int> s ; 

int main(){
    long long  n ; 
    cin >> n ; 
    while(n > 0){
        long long arr[n+1] ;
        long long  maxArea  = 0 ; 
        long long currentArea = 0 ;
        long long  i = 0 ; 
        for(i = 0 ; i < n ; i++)
            cin >> arr[i];

        // Processing 
        i = 0 ; 
        while(i < n){
            if(s.empty() || arr[i] >= arr[s.top()]){
                s.push(i++);
            }

            else{
                long long  temp = s.top() ; 
                s.pop() ;
                currentArea = arr[temp] * (s.empty() ? i : i - s.top() - 1);
                maxArea = max(maxArea, currentArea);
            }
        }

    
        // Pop the remaining elements from the stack
        while(! s.empty()){
            long long temp = s.top() ;
            s.pop() ;
            currentArea = arr[temp] * (s.empty() ? i : i - s.top() - 1);
            maxArea = max(maxArea, currentArea);
        }
 
        // Output
        cout << maxArea << endl ;
        cin >> n ;
        
    } 

return  0 ;
}
